proposals/0052-svg-components.md

```diff
@@ -0,0 +1,293 @@
+
+# Non-Goals
+
+- Out-of-scope: Advanced tree-shaking or SSR optimizations for large icon sets.
+- Out-of-scope: Advanced `.svg` file optimization such as `svgo`.
+- Future: Adding support for `.svg` files as framework components.
+- Non-goal: Adding support for `.json` icon sets such as `@iconify/json`.
+- Non-goal: SVG content analysis for accessibility.
+
+# Detailed Design
+
+## Importing SVG Files as Components
+
+The core design of this proposal allows `.svg` files to be imported and used as components in Astro. This means developers can import `.svg` files into their Astro components and render them directly as if they were Astro components. The goal is to maintain ease of use while automating performance optimizations under the hood.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg';
+---
+
+<Logo />
+```
+This would generate the following output:
+
+```astro
+<svg width="24" height="24" role="img">
+    <!-- SVG Content -->
+</svg> 
+```
+
+## Render Methods
+
+There are two ways that the SVG component can render the `.svg` file, inline or as a sprite.
+
+### Inline
+
+Each time a `.svg` file is imported and rendered in Astro, Astro will directly inject the optimized `.svg` file contents into the location with the root `<svg>` element properties overridden by the props passed to the SVG Component.
+
+This follows the most common implementation of rendering SVGs into the HTML. This method is the safest option as it does the least manipulation to the `.svg` file. However, the tradeoff is that repeated usages of the same `.svg` file will result in multiple copies in the HTML.
+
+**Example:**
+
+```astro
+---
+import Logo from '../assets/logo.svg';
+---
+
+<Logo class="text-slate-500" />
+```
+
+This would generate the following output:
+
+```astro
+<svg width="24" height="24" role="img" class="text-slate-500" >
+    <!-- SVG Content -->
+</svg> 
+```
+
+### Sprite
+
+An SVG Sprite is when an `<svg>` element defines multiple SVGs (or symbols) using the `<symbol>` element and attaching a unique identifier to each SVG as an `id`. These are then able to be referenced using the `<use>` element with an `href` attribute that points to the `id` defined on the `<symbol>` element. 
+
+When a `.svg` file is first imported and rendered in Astro, the system will convert the file into a `<symbol>` element. This will be inserted into the `<svg>` element. This approach ensures that all subsequent renders of that `.svg` can use a `<use>` element, referencing the ID of the initial `<symbol>`.
+
+This method was chosen by `astro-icon` to optimize SVG usage by default. It is riskier because it does more manipulation to the `.svg` file and relies on the `id` reference being available. The benefit to this method is that we now only have a single definition of repeated usages of the same `.svg` file and only minimal code to reference the definition in other locations.
```

Comment:
This is opt-in only when the user uses `<symbol>`, is that right?

Reply:
This is opt-in through a prop on the icon or at the config level.

Reply:
I was wanting to also investigate creating a `Sprite` component that would hold all the definitions. This could solve the edge-case of undefined icons. Though I'm not sure if it's possible.